### ISSUE TITLE
OnSysColourChanged during system lock/unlock changes BG color

### ISSUE THREAD

**Issue by sz5000** [wxWidgets/wxWidgets#23386]
In the function `wxToolBar::OnSysColourChanged(wxSysColourChangedEvent& event)` of the Windows port (src/msw/toolbar.cpp, line 1955) `m_backgroundColour` is updated to `::GetSysColor(COLOR_BTNFACE)`. This happens on my system when I lock my system (WINDOWS + L) and unlock it afterwards. 

The problem now is the following:

If I change the background color in my app to something other than the default color and lock my system and then unlock it again, the background color is changed and the toolbar doesn't look the same. Isn't it better to only change the background color if it hasn't been changed by the app?

As a workaround, we can implement an event handler for `wxSysColourChangedEvent` to update the color again, but to me it seems more like a problem than a feature.

Maybe someone here can say more about this behavior.

**Comment by vadz** [comment]
Can the problem be seen in the unmodified toolbar sample?

**Comment by sz5000** [comment]
Yes, I've compile the toolbar sample and tested with that, same happens here.

![image](https://user-images.githubusercontent.com/33447587/228243002-e7a19b51-e654-43b7-a867-a16368afcf3c.png)

**Comment by vadz** [comment]
Sorry, just to be totally clear: you've add a `SetBackgroundColour()` call to the sample? I'll try to test this myself, but it's really nice to have a patch (even if a trivial one) allowing to reproduce the problem at hand.

**Comment by sz5000** [comment]
You do not need to change anything.  The color can be changed via the entry in the menu bar.

![image](https://user-images.githubusercontent.com/33447587/228250464-51b003b4-847b-472f-9c97-f47d1e3a07bf.png)

**Comment by sz5000** [comment]
Is the following line possibly simply unnecessary?

`wxRGBToColour(m_backgroundColour, ::GetSysColor(COLOR_BTNFACE));`

**Comment by vadz** [comment]
> Is the following line possibly simply unnecessary?
> 
> `wxRGBToColour(m_backgroundColour, ::GetSysColor(COLOR_BTNFACE));`

This is needed, of course, when using the default background colour -- otherwise it wouldn't update when the system colours really change.

I think this simple change:
```diff
diff --git a/src/msw/toolbar.cpp b/src/msw/toolbar.cpp
index 5050d3a4a9..106f13fa12 100644
--- a/src/msw/toolbar.cpp
+++ b/src/msw/toolbar.cpp
@@ -1954,7 +1954,8 @@ void wxToolBar::SetToolPacking(int packing)
 // Responds to colour changes, and passes event on to children.
 void wxToolBar::OnSysColourChanged(wxSysColourChangedEvent& event)
 {
-    wxRGBToColour(m_backgroundColour, ::GetSysColor(COLOR_BTNFACE));
+    if ( !UseBgCol() )
+        wxRGBToColour(m_backgroundColour, ::GetSysColor(COLOR_BTNFACE));
 
     // Remap the buttons
     Realize();
```
would be enough to fix the problem you're seeing, but it's weird that I don't see at all because no `WM_SYSCOLORCHANGE` is generated at all when locking/unlocking Windows (tested with 10.0.19043.2251). I have no idea why does it result in this even for you, but this seems wrong. Could you please put a breakpoint on this function in the debugger and check where is it called from when this happens? You'd need to configure using Microsoft symbol servers to see anything useful, of course.

**Comment by sz5000** [comment]
Hard to say where the message comes from. Seems to come from outside. Maybe someone is sending the message to all top-level windows after log on. I also use Windows 10 22H2 19045.2788.
